demos/induction_heads_multigpu.py

```diff
@@ -0,0 +1,285 @@
+            device=torch.cuda.current_device(),
+            recurse=False,
+        )
+
+    auto_wrap_policy = partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={
+            LlamaDecoderLayer,
+        },
+    )
+    sharding_strategy = ShardingStrategy.FULL_SHARD
+    device_id = torch.cuda.current_device()
+    sync_module_states = True
+    param_init_fn = _module_init_fn if rank != 0 else None
+    mp_policy = MixedPrecision(
+        param_dtype=torch.bfloat16,
+        buffer_dtype=torch.bfloat16,
+        reduce_dtype=torch.bfloat16,
+    )
+
+    config = {
+        "auto_wrap_policy": auto_wrap_policy,
+        "sharding_strategy": sharding_strategy,
+        "device_id": device_id,
+        "sync_module_states": sync_module_states,
+        "param_init_fn": param_init_fn,
+        "mixed_precision": mp_policy,
+    }
+    return config
+
+
+def calculate_induction_score(
+        config: LlamaConfig,
+        activation_dict: dict[str, torch.Tensor],
+        module_names: list[str],
+        sequence_length: int,
+    ) -> None:
+    """Calculate and save a heatmap of the induction scores for each attention
+    head.
+
+    Args:
+    ----
+        config: The model's HF config
+        activation_dict: Dictionary containing the activations retrieved using
+            FlexModel
+        module_names: A list of the module names to which we have attached
+            hooks
+        sequence_length: The sequence length of the prompt passed into the
+            model
+    """
+    induction_score_store = torch.zeros(
+        (
+            config.num_hidden_layers,
+            config.num_attention_heads,
+        ),
+        device=torch.cuda.current_device(),
+    )
+
+    for i, module_name in enumerate(module_names):
+        attn_maps = activation_dict[module_name][0].detach().to(
+            torch.cuda.current_device(),
+        )
+        induction_stripe = attn_maps.diagonal(
+            dim1=-2, dim2=-1, offset=1-sequence_length,
+        )
+        induction_score = einops.reduce(
+            induction_stripe, "batch head_index position -> head_index", "mean",
+        )
+        induction_score_store[i, :] = induction_score
+
+    plt.imshow(induction_score_store.cpu().numpy(), origin="lower")
+    plt.xlabel("Head")
+    plt.ylabel("Layer")
+    plt.title("Induction Score by Head")
+    plt.colorbar()
+    plt.savefig("induction_score_by_head.png", bbox_inches="tight")
+
+
+def get_module_names(config: LlamaConfig) -> list[str]:
+    """Return the list of module names to apply hooks onto.
+
+    Args:
+    ----
+        config: The model's config
+
+    Returns:
+    -------
+        A list of model names that we're applying HookFunctions to
+    """
+    name_placeholder = """_fsdp_wrapped_module.model.layers.
+        {}._fsdp_wrapped_module.self_attn.dummy"""
+    module_names = [
+        name_placeholder.format(i) for i in range(config.num_hidden_layers)
+    ]
```

Comment:
Use f-string instead here, ie. `module_names = [f"layer_prefix.{i}.layer_postfix" for i in range(...)]`